### PR TITLE
Significant disambiguation between host and target.

### PR DESCRIPTION
**Pull request by jaykrell**
mgreg_t often to vs. host_mgreg_t to fix C++ cross compilation.

This is much noiser alternative to https://github.com/mono/mono/pull/10881

### CODE REVIEW

File: mono/mini/exceptions-amd64.c [hunk 1]
```diff
@@ -486,47 +498,47 @@ get_throw_trampoline (MonoTrampInfo **info, gboolean rethrow, gboolean corlib, g
...
+	arg_offsets [2] = dummy_stack_space + sizeof(host_mgreg_t) * 2;
+	arg_offsets [3] = dummy_stack_space + sizeof(host_mgreg_t) * 3;
+	ctx_offset = dummy_stack_space + sizeof(host_mgreg_t) * 4;
 	regs_offset = ctx_offset + MONO_STRUCT_OFFSET (MonoContext, gregs);
 
 	/* Save registers */
 	for (i = 0; i < AMD64_NREG; ++i)
 		if (i != AMD64_RSP)
-			amd64_mov_membase_reg (code, AMD64_RSP, regs_offset + (i * sizeof(mgreg_t)), i, sizeof(mgreg_t));
+			amd64_mov_membase_reg (code, AMD64_RSP, regs_offset + (i * sizeof(host_mgreg_t)), i, sizeof(host_mgreg_t));
```
Comment (vargaz) [inline mono/mini/exceptions-amd64.c]: The generated code in general should use target_mgreg_t.

**Comment by jaykrell** [comment]
How about  all  the `mono_arch_get_throw_`? I'll dial those back.

**Comment by jaykrell** [comment]
Oh I see. Those were correct.
```
	if (mono_ee_features.use_aot_trampolines) {
		restore_context_func = mono_aot_get_trampoline ("restore_context");
		call_filter_func = mono_aot_get_trampoline ("call_filter");
		throw_exception_func = mono_aot_get_trampoline ("throw_exception");
		rethrow_exception_func = mono_aot_get_trampoline ("rethrow_exception");
		rethrow_preserve_exception_func = mono_aot_get_trampoline ("rethrow_preserve_exception");
	} else {
		MonoTrampInfo *info;

		restore_context_func = mono_arch_get_restore_context (&info, FALSE);
		mono_tramp_info_register (info, NULL);
		call_filter_func = mono_arch_get_call_filter (&info, FALSE);
		mono_tramp_info_register (info, NULL);
		throw_exception_func = mono_arch_get_throw_exception (&info, FALSE);
		mono_tramp_info_register (info, NULL);
		rethrow_exception_func = mono_arch_get_rethrow_exception (&info, FALSE);
		mono_tramp_info_register (info, NULL);
		rethrow_preserve_exception_func = mono_arch_get_rethrow_preserve_exception (&info, FALSE);
		mono_tramp_info_register (info, NULL);
	}
```

**Comment by jaykrell** [comment]
@monojenkins build Linux i386

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 C++

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins build Linux WebAssembly

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 FullAOT+LLVM

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/mini/exceptions-amd64.c [hunk 2]
```diff
@@ -919,6 +949,8 @@ mono_arch_handle_altstack_exception (void *sigctx, MONO_SIG_HANDLER_INFO_TYPE *s
 GSList*
 mono_amd64_get_exception_trampolines (gboolean aot)
 {
+	mono_cross_compile_assert_not_reached ();
+
```
Comment (vargaz) [inline mono/mini/exceptions-amd64.c]: This is incorrect, and probably a bunch of others as well.

Comment (jaykrell) [reply mono/mini/exceptions-amd64.c]: I think I understand and will make another pass.

**Comment by vargaz** [comment]
The mgreg_t changes look ok to me, the cross_compile_assert changes look incorrect in many places, I think they should only be added to a limited set of functions which deal with host_mgreg_t etc.

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by jaykrell** [comment]
@monojenkins Linux x64

**Comment by jaykrell** [comment]
@monojenkins build Windows x64 C++

**Comment by jaykrell** [comment]
@monojenkins build Windows i386

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build Linux i386

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv7

**Comment by jaykrell** [comment]
@monojenkins Linux ARMv5

**Comment by jaykrell** [comment]
@monojenkins Linux x64 Coop suspend

**Comment by jaykrell** [comment]
@monojenkins  build linux ARMv5

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv5

**Comment by jaykrell** [comment]
@monojenkins build  Linux x64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Coop Suspend

**Comment by jaykrell** [comment]
Good now? I guess I’ll rebase to master to help CI go green.

**Comment by jaykrell** [comment]
@monojenkins Linux x64 Coop Suspend
@monojenkins Linux x64 Interpreter
@monojenkins Linux x64 WebAssembly
@monojenkins Linux x64 WebAssembly C++

**Comment by jaykrell** [comment]
@monojenkins Linux x64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins Linux x64 Interpreter

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by jaykrell** [comment]
All 3 sdk C++ pass.

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by jaykrell** [comment]
Linux x64 Interpreter is known and unrelated.

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Interpreter

**Pull request merged**